### PR TITLE
Add greedy search and modified_beam_search for non-streaming ASR

### PR DESCRIPTION
**Pull request by csukuangfj**

### CODE REVIEW

**Comment by csukuangfj** [comment]
This PR is ready for review.

# Usage
## Installation
```bash
pip install kaldifeat kaldi_native_io
git clone https://github.com/k2-fsa/sherpa
cd sherpa
mkdir build
cd build
cmake ..
make -j sherpa
```

You will find an executable `bin/sherpa` inside the build directory.

```bash
./bin/sherpa --help
```
will show you the usage information.


**Note**: We will eventually support `conda install`.

## Download the pre-trained model
```bash
git lfs install
git clone https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13
```

## Decode a single wave
```bash
./build/bin/sherpa \
  --decoding-method=greedy_search \
  --nn-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/exp/cpu_jit-torch-${{ matrix.torch }}.pt \
  --bpe-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/data/lang_bpe_500/bpe.model \
  icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/test_wavs/1089-134686-0001.wav
```

**Note**: You can also use `--decoding-method=modified_beam_search`.

## Decoding multiple waves
```bash
./build/bin/sherpa \
  --decoding-method=greedy_search \
  --nn-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/exp/cpu_jit-torch-${{ matrix.torch }}.pt \
  --bpe-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/data/lang_bpe_500/bpe.model \
  icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/test_wavs/1089-134686-0001.wav \
  icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/test_wavs/1221-135766-0001.wav \
  icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/test_wavs/1221-135766-0002.wav
```

## Decode wav.scp
```bash
./build/bin/sherpa \
  --decoding-method=greedy_search \
  --nn-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/exp/cpu_jit-torch-${{ matrix.torch }}.pt \
  --bpe-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/data/lang_bpe_500/bpe.model \
  --use-wav-scp=true \
  scp:wav.scp \
  ark,scp,t:results.ark,results.scp
```

## Decode feats.scp
```bash
./build/bin/sherpa \
  --decoding-method=greedy_search \
  --nn-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/exp/cpu_jit-torch-${{ matrix.torch }}.pt \
  --bpe-model=./icefall-asr-librispeech-pruned-transducer-stateless3-2022-05-13/data/lang_bpe_500/bpe.model \
  --use-feats-scp=true \
  scp:feats.scp \
  ark,scp,t:results3.ark,results3.scp
```

One thing to note for decoding `feats.scp`: Kaldi uses samples in the range [-32768, 32767] to compute the features while
models from icefall uses samples in the range [-1, 1) to compute the features, so you cannot use models from icefall
to decode `feats.scp` generated by kaldi's `compute-fbank-feats`.

You can use the following file to generate `feats.scp` and `feats.ark` for testing
https://github.com/k2-fsa/sherpa/blob/a92b5903bcd63813f035ece873b24f5ebf86e71d/.github/scripts/generate_feats_scp.py

https://github.com/k2-fsa/sherpa/blob/a92b5903bcd63813f035ece873b24f5ebf86e71d/.github/scripts/generate_feats_scp.py#L1-L29

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,6 +48,8 @@ include(pybind11)
 include(torch)
 include(k2)
 include(kaldifeat)
+include(kaldi_native_io)
+include(sentencepiece)
```
Comment (pkufool) [inline CMakeLists.txt:52]: I don't think we really need sentencepiece in decoding, we always generate tokens for the bpe model. I suggest to pass a argument like `--use-bpe`, if it is True then we can replace the "▁" with " " in the recognized texts.

**Review by pkufool — APPROVED** [review]
LGTM

File: CMakeLists.txt (hunk 1 above)
Comment (csukuangfj) [reply CMakeLists.txt]: Yes, you are right. Fixed.

**Pull request merged**